### PR TITLE
RES-112 Implement Sequence ID

### LINKED ISSUE #2741: Partition (sequence) ID support
Github issue corresponding to https://jira.numenta.com/browse/RES-112

We want to be able to pass the sequence ID info throughout the network; this is useful for debugging purposes. The [partitionID in the KNNClassifier|https://github.com/numenta/nupic/blob/master/src/nupic/algorithms/KNNClassifier.py#L359] is supposed to do this, but it's not clear this works. So the task is to first test to see if this is functioning properly. Then we need to pass the sequence (partition) ID number throughout the network. The LanguageSensor already has an output 

["sequenceIdOut"|https://github.com/numenta/nupic.research/blob/master/htmresearch/regions/LanguageSensor.py#L101], but this needs to be linked to the subsequent regions in the network. This requires (i) that these regions (SP, TM, TP, KNN, and CLA) have inputs for "sequenceIdIn", and (ii) that the links are created by the network factory...
i. add to the the spec:
          sequenceIdIn=dict(
            description="Sequence ID",
            dataType='UInt64',
            count=1,
            required=False,
            regionLevel=True,
            isDefaultInput=False,
            requireSplitterMap=False)
ii. add the following in the network factory's linking method:
  network.link(sensorRegionName, currentRegionName, "UniformLink", "",
               srcOutput="sequenceIdOut", destInput="sequenceIdIn")

### PR DESCRIPTION
**Pull request by SaganBolliger**
https://jira.numenta.com/browse/RES-112 (fixes #2741).

This PR reopens #2726 as decided in discussions with @subutai and @BoltzmannBrain.

### CODE REVIEW

**Comment by rhyolight** [comment]
@SaganBolliger Please ping someone to review this.

**Comment by BoltzmannBrain** [comment]
@SaganBolliger there needs to be a corresponding link made in the [classification network factory](https://github.com/numenta/nupic.research/blob/master/htmresearch/frameworks/classification/classification_network.py#L193).

**Comment by SaganBolliger** [comment]
@BoltzmannBrain I've created PR numenta/nupic.research#370 to address this. Does this look go to you?

**Comment by BoltzmannBrain** [comment]
@SaganBolliger no, please address my comments in https://github.com/numenta/nupic.research/pull/370. You've setup the inputs here, but in the network factory in nupic.research there is nothing that links them to the sensor.

**Comment by BoltzmannBrain** [comment]
@SaganBolliger it looks like some tests need to be updated, i.e. `tests.integration.nupic.opf.opf_region_test.OPFRegionTest.testGetInputOutputNamesOnRegions` fails the CI build

**Comment by SaganBolliger** [comment]
@BoltzmannBrain I've updated the integration test you mentioned. I also made the test invariant to the order of the inputs/outputs by doing set comparison. This was also causing the test to fail, and the order doesn't have any meaning, from what I understand.

Do you know why that integration test isn't being run by Travis CI, but is by AV?

**Comment by BoltzmannBrain** [comment]
:+1: 
@SaganBolliger I'm unsure, that's a question for @rhyolight

**Comment by BoltzmannBrain** [comment]
@subutai please merge (nupic.research [#370](https://github.com/numenta/nupic.research/pull/370) relies on this)

File: src/nupic/regions/KNNClassifierRegion.py [hunk 1]
```diff
@@ -84,7 +84,17 @@ def getSpec(cls):
             required=False,
             regionLevel=True,
             isDefaultInput=False,
-            requireSplitterMap=False)
+            requireSplitterMap=False),
+
+          sequenceIdIn=dict(
```
Comment (subutai) [inline src/nupic/regions/KNNClassifierRegion.py]: As discussed, please remove sequenceIdIn from KNNClassifierRegion and then I can merge. Thanks!

**Pull request merged**